### PR TITLE
Cut down on the number of AccessTraits<Predicates, PredicatesTag>

### PR DESCRIPTION
**Pull request by aprokop**
Follow up to #967.

### CODE REVIEW

**Comment by aprokop** [comment]
After spending quite a bit of time on it, I have no idea how to fix the CUDA 11.0.3 `ptxas fatal` error. Def a compiler bug, but can't figure out how to work around it.

**Comment by masterleinad** [comment]
It seems to complain about unresolved extern functions for
```
ArborX::AccessTraits<Kokkos::View<ArborX::Intersects<ArborX::Box>*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> >, ArborX::PredicatesTag, void>::get(Kokkos::View<ArborX::Intersects<ArborX::Box>*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> > const&, int)
ArborX::Intersects<ArborX::ExperimentalHyperGeometry::Sphere<3, float> >&& std::forward<ArborX::Intersects<ArborX::ExperimentalHyperGeometry::Sphere<3, float> > >(std::remove_reference<ArborX::Intersects<ArborX::ExperimentalHyperGeometry::Sphere<3, float> > >::type&)
ArborX::AccessTraits<Kokkos::View<ArborX::Intersects<ArborX::Sphere>*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> >, ArborX::PredicatesTag, void>::get(Kokkos::View<ArborX::Intersects<ArborX::Sphere>*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace> > const&, int)
```

File: src/ArborX_DistributedTree.hpp [hunk 1]
```diff
@@ -86,16 +87,24 @@ class DistributedTree
...
+    static_assert(
+        KokkosExt::is_accessible_from<typename Predicates::memory_space,
+                                      ExecutionSpace>::value,
+        "Predicates must be accessible from the execution space");
+
+    using Tag = typename Predicates::value_type::Tag;
     using DeviceType = Kokkos::Device<ExecutionSpace, MemorySpace>;
     Details::DistributedTreeImpl<DeviceType>::queryDispatch(
-        Tag{}, *this, space, predicates, std::forward<Args>(args)...);
+        Tag{}, *this, space, Predicates{user_predicates},
```
Comment (dalg24) [inline src/ArborX_DistributedTree.hpp:106]: Nitpick: just define predicates before dispatching.

File: src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp [hunk 2]
```diff
@@ -150,9 +146,8 @@ void queryImpl(ExecutionSpace const &space, Tree const &tree,
   {
     tree.query(
         space, permuted_predicates,
-        InsertGenerator<FirstPassTag, PermutedPredicates, Callback, OutputView,
-                        CountView, PermutedOffset>{callback, out, counts,
-                                                   permuted_offset},
+        InsertGenerator<FirstPassTag, Callback, OutputView, CountView,
```
Comment (dalg24) [inline src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp:149]: Do you need to specify the template parameters other than the tag?

Comment (aprokop) [reply src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp:149]: I tried removing them, and it failed. I always forget what the rules are.

File: src/ArborX_DistributedTree.hpp (hunk 1 above)
Comment (aprokop) [reply src/ArborX_DistributedTree.hpp:106]: I don't understand. You want `Predicates predicates{user_predicates};` before `queryDispatch`?

**Comment by aprokop** [comment]
Rebased on master to resolve conflicts with #970.

File: src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp (hunk 2 above)
Comment (masterleinad) [reply src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp:149]: If one template parameter is specified, all of them need to be provided, see https://godbolt.org/z/eKhnTW3nr. Aggregate deduction only works if no template parameters are given (which would require storing the tag here).

File: src/ArborX_DistributedTree.hpp (hunk 1 above)
Comment (dalg24) [reply src/ArborX_DistributedTree.hpp:106]: yes

File: src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp (hunk 2 above)
Comment (aprokop) [reply src/details/ArborX_DetailsCrsGraphWrapperImpl.hpp:149]: Thank you.

File: src/ArborX_DistributedTree.hpp (hunk 1 above)
Comment (aprokop) [reply src/ArborX_DistributedTree.hpp:106]: Do you want the same in BVH and BruteForce? It's the same code.

Comment (dalg24) [reply src/ArborX_DistributedTree.hpp:106]: Up to you. I mostly didn't like the formatting change here.

**Review by dalg24 — APPROVED** [review]
Good enough

**Comment by aprokop** [comment]
> Good enough

Yeah. Except it f-ing does not build on CUDA-11.0.3 because of `ptxas` and I have no idea how to fix that.

File: src/details/ArborX_DetailsPermutedData.hpp [hunk 3]
```diff
@@ -23,9 +23,17 @@ namespace Details
...
+
   Data _data;
   Permute _permute;
-  KOKKOS_FUNCTION auto &operator()(int i) const { return _data(_permute(i)); }
+
+  KOKKOS_FUNCTION decltype(auto) operator()(int i) const
+  {
+    return _data(_permute(i));
+  }
+  KOKKOS_FUNCTION decltype(auto) size() const { return _data.size(); }
```
Comment (dalg24) [inline src/details/ArborX_DetailsPermutedData.hpp:36]: No point to decltype here

**Comment by aprokop** [comment]
Given that I could not figure out how to fix cuda/11.0.3 build with nvcc, we decided to (hopefully temporarily) switch the cuda version in the CI to cuda/11.1 to make progress with refactoring. The hope is that after we complete the refactoring, we would try to switch back and maybe the issue would disappear. The goal is to not waste time on it right now. 

Having said that, we would welcome any attempts to fix it.

**Comment by aprokop** [comment]
CUDA 11.1.1 passes. SYCL is broken until #993 or #994 is fixed.

**Pull request merged**